Sources/NIO/Selector.swift

```diff
@@ -540,7 +543,8 @@ final class Selector<R: Registration> {
             Int(try KQueue.kevent(kq: self.fd, changelist: nil, nchanges: 0, eventlist: events, nevents: Int32(eventsCapacity), timeout: ts))
         }
 
-        for i in 0..<ready {
+        self.deregistrationsHappened = false
+        for i in 0..<ready where !self.deregistrationsHappened {
```

Comment:
as noted offline I think while this should work we should treat this as workaround as this depends highly on the "level-triggered" semantics of kqueue / epoll.

Reply:
Totally agreed. Let’s get this in (I actually found another bug (and a missing assertion) whilst fixing it) and then fix properly. I agree we shouldn’t rely on it being level triggered here.

Reply:
Works for me